core/blockchain_test.go

```diff
@@ -4326,3 +4332,75 @@ func TestEIP3651(t *testing.T) {
 		t.Fatalf("sender balance incorrect: expected %d, got %d", expected, actual)
 	}
 }
+
+func TestRewindOnConfigChange(t *testing.T) {
+	genesisTime := uint64(12)
+
+	type testCase struct {
+		name              string
+		override1         func(*params.ChainConfig)
+		override2         func(*params.ChainConfig)
+		expectChainRewind bool
+	}
+
+	tcs := []testCase{
+		{
+			name:              fmt.Sprintf("CanyonTime changes from 10 to 0 (genesis time is %d)", genesisTime),
+			override1:         func(c *params.ChainConfig) { c.CanyonTime = uint64ptr(10) },
+			override2:         func(c *params.ChainConfig) { c.CanyonTime = uint64ptr(0) },
+			expectChainRewind: false,
+		},
+		{
+			name:              fmt.Sprintf("RegolithTime changes from 10 to 0 (genesis time is %d)", genesisTime),
+			override1:         func(c *params.ChainConfig) { c.RegolithTime = uint64ptr(10) },
+			override2:         func(c *params.ChainConfig) { c.RegolithTime = uint64ptr(0) },
+			expectChainRewind: false,
+		},
+		{
+			name:              fmt.Sprintf("ShanghaiTime changes from 10 to 0 (genesis time is %d)", genesisTime),
+			override1:         func(c *params.ChainConfig) { c.ShanghaiTime = uint64ptr(10) },
+			override2:         func(c *params.ChainConfig) { c.ShanghaiTime = uint64ptr(0) },
+			expectChainRewind: false,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			var genesis = &Genesis{
+				BaseFee:   big.NewInt(params.InitialBaseFee),
+				Config:    params.AllEthashProtocolChanges,
+				Timestamp: genesisTime,
+			}
+
+			// Prepare initial config for chain
+			tc.override1(genesis.Config)
+			db := rawdb.NewMemoryDatabase()
+			cc := DefaultCacheConfigWithScheme(rawdb.PathScheme)
+
+			// Start blockchain once to store config in DB
+			blockchain, _ := NewBlockChain(db, cc, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
+
+			// Stop chain after 1 second
+			<-time.After(1 * time.Second)
```

Comment:
There's `time.Sleep`. And is there some polling to the `blockchain` that we can do to confirm the db has been written to once, that we could poll in a loop with a shorter timeout (possibly using `require.Eventually`)? This would be quicker and at the same time less flaky than assuming that after 1sec the desired state has been achieved.